aws_xray_sdk/core/plugins/ec2_plugin.py

```diff
@@ -17,16 +19,51 @@ def initialize():
     """
     global runtime_context
 
+    # Try the IMDSv2 endpoint for metadata
     try:
         runtime_context = {}
 
-        r = urlopen('http://169.254.169.254/latest/meta-data/instance-id', timeout=1)
-        runtime_context['instance_id'] = r.read().decode('utf-8')
+        # get session token with 60 seconds TTL to not have the token lying around for a long time
+        token = do_request(url=IMDS_URL + "api/token",
+                           headers={"X-aws-ec2-metadata-token-ttl-seconds": "60"},
+                           method="PUT")
+
+        # get instance-id metadata
+        runtime_context['instance_id'] = do_request(url=IMDS_URL + "meta-data/instance-id",
+                                                    headers={"X-aws-ec2-metadata-token": token},
+                                                    method="GET")
+
+        # get availability-zone metadata
+        runtime_context['availability_zone'] = do_request(url=IMDS_URL + "meta-data/placement/availability-zone",
+                                                          headers={"X-aws-ec2-metadata-token": token},
+                                                          method="GET")
+
+    except Exception as e:
+        # Falling back to IMDSv1 endpoint
+        log.warning("failed to get ec2 instance metadata from IMDSv2 due to {}. Falling back to IMDSv1".format(e))
+
+        try:
+            runtime_context = {}
 
-        r = urlopen('http://169.254.169.254/latest/meta-data/placement/availability-zone',
-                    timeout=1)
-        runtime_context['availability_zone'] = r.read().decode('utf-8')
+            runtime_context['instance_id'] = do_request(url=IMDS_URL + "meta-data/instance-id")
 
+            runtime_context['availability_zone'] = do_request(url=IMDS_URL + "meta-data/placement/availability-zone-1")
+
+        except Exception as e:
+            runtime_context = None
+            log.warning("failed to get ec2 instance metadata from IMDSv1 due to {}".format(e))
+
+
+def do_request(url, headers=None, method="GET"):
+    if headers is None:
+        headers = {}
+    try:
+        if url is None:
+            return None
+        req = Request(url=url)
+        req.headers = headers
+        req.method = method
+        res = urlopen(req, timeout=1)
+        return res.read().decode('utf-8')
     except Exception:
-        runtime_context = None
-        log.warning("failed to get ec2 instance metadata.")
+        raise
```

Comment:
If we're just raising these exceptions anyway does it need to be in try/except block?

Reply:
You're right. Don't really need the try/except block

Reply:
Correct. No need for the try/except block here.

Reply:
removed the try except block